toolchains/private/collect.bzl

```diff
@@ -0,0 +1,118 @@
+# Copyright 2025 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper functions to allow us to collect data from attr.label_list."""
+
+load(
+    "//toolchains:toolchain_info.bzl",
+    "ActionTypeSetInfo",
+    "ArgsListInfo",
+    "FeatureSetInfo",
+    "ToolInfo",
+)
+
+visibility([
+    "//toolchains/...",
+    "//tests/...",
+])
+
+def _make_collector(provider, field):
+    def collector(targets, direct = [], transitive = []):
+        # Avoid mutating what was passed in.
+        transitive = transitive[:]
+        for target in targets:
+            transitive.append(getattr(target[provider], field))
+        return depset(direct = direct, transitive = transitive)
+
+    return collector
+
+collect_action_types = _make_collector(ActionTypeSetInfo, "actions")
+
+def collect_action_labels(actions):
+    return [action.label for action in collect_action_types(actions).to_list()]
+
+collect_files = _make_collector(DefaultInfo, "files")
+
+def collect_data(targets):
+    """Collects from a 'data' attribute.
+
+    This is distinguished from collect_files by the fact that data attributes
+    attributes include runfiles.
+
+    Args:
+        targets: (List[Target]) A list of files or executables
+
+    Returns:
+        A list of FilesToRunProvider and depset[file] containing the files required
+        required to run them.
+    """
+    out = []
+    for target in targets:
+        info = target[DefaultInfo]
+        if info.default_runfiles != None:
+            out.append(info.default_runfiles.files)
```

Comment:
This will lose the symlinks. Instead, we should use `ctx.runfiles().merge_all()` and return a `runfiles` object. The comment needs to be updated.